{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "773855f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from Bio import SeqIO\n",
    "\n",
    "def read_data():\n",
    "    base_directory = './e_coli_paidb/'\n",
    "    data = {}\n",
    "    \n",
    "    for name in os.listdir(base_directory):\n",
    "        dir_path = os.path.join(base_directory, name)\n",
    "\n",
    "        if os.path.isdir(dir_path):\n",
    "            fasta_file = os.path.join(dir_path, f\"{name}.fasta\")\n",
    "            pai_file   = os.path.join(dir_path, f\"{name}.pai\")\n",
    "            cpai_file  = os.path.join(dir_path, f\"{name}.cpai\")\n",
    "            npai_file  = os.path.join(dir_path, f\"{name}.npai\")\n",
    "\n",
    "            data[name] = {}\n",
    "            genome_sequence = next(SeqIO.parse(fasta_file, 'fasta'))\n",
    "            data[name]['genome_sequence'] = genome_sequence\n",
    "\n",
    "            with open(pai_file, 'r') as file:\n",
    "                data[name]['pai_sequences'] = []\n",
    "\n",
    "                while True:\n",
    "                    line = file.readline()\n",
    "                    if not line:\n",
    "                        break\n",
    "\n",
    "                    line = line.split()\n",
    "                    pai_start_index = int(line[0])\n",
    "                    pai_end_index = int(line[1])\n",
    "\n",
    "                    pai_sequence = genome_sequence.seq[pai_start_index:pai_end_index]\n",
    "\n",
    "                    data[name]['pai_sequences'].append((pai_sequence, (pai_start_index, pai_end_index)))\n",
    "\n",
    "            with open(cpai_file, 'r') as file:\n",
    "                data[name]['cpai_sequences'] = []\n",
    "\n",
    "                while True:\n",
    "                    line = file.readline()\n",
    "                    if not line:\n",
    "                        break\n",
    "\n",
    "                    line = line.split()\n",
    "                    cpai_start_index = int(line[0])\n",
    "                    cpai_end_index = int(line[1])\n",
    "\n",
    "                    cpai_sequence = genome_sequence.seq[cpai_start_index:cpai_end_index]\n",
    "\n",
    "                    data[name]['cpai_sequences'].append((cpai_sequence, (cpai_start_index, cpai_end_index)))\n",
    "\n",
    "            with open(npai_file, 'r') as file:\n",
    "                data[name]['npai_sequences'] = []\n",
    "\n",
    "                while True:\n",
    "                    line = file.readline()\n",
    "                    if not line:\n",
    "                        break\n",
    "\n",
    "                    line = line.split()\n",
    "                    npai_start_index = int(line[0])\n",
    "                    npai_end_index = int(line[1])\n",
    "\n",
    "                    npai_sequence = genome_sequence.seq[npai_start_index:npai_end_index]\n",
    "\n",
    "                    data[name]['npai_sequences'].append((npai_sequence, (npai_start_index, npai_end_index)))\n",
    "\n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "47f0fe7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "\n",
    "def tfidf(sequence, k):\n",
    "    kmers = get_kmers(str(sequence), k)\n",
    "    vectorizer = TfidfVectorizer()\n",
    "    X = vectorizer.fit_transform([kmers])\n",
    "    tfidf_matrix = X.toarray()\n",
    "    return tfidf_matrix, vectorizer\n",
    "\n",
    "def get_kmers(sequence, k):\n",
    "    kmers = [sequence[i:i+k] for i in range(len(sequence) - k + 1)]\n",
    "    return ' '.join(kmers)\n",
    "\n",
    "def get_kmers_map(tfidf_matrix, vectorizer, sequence, top_n=20):\n",
    "    feature_names = vectorizer.get_feature_names_out()\n",
    "    feature_scores = tfidf_matrix.flatten()\n",
    "    \n",
    "    top_kmers_indices = feature_scores.argsort()[-top_n:][::-1]\n",
    "    top_kmers = [(feature_names[i], feature_scores[i]) for i in top_kmers_indices]\n",
    "\n",
    "    mapped_kmers = {}\n",
    "    for kmer, score in top_kmers:\n",
    "        kmer = kmer.upper()\n",
    "        count = sequence.count(kmer)\n",
    "        mapped_kmers[kmer] = {'score': score, 'count': count}\n",
    "        \n",
    "    return mapped_kmers\n",
    "\n",
    "def find_unique_kmers(full_sequence, start_index, end_index, significant_kmers, unique_kmers):\n",
    "    for kmer, details in significant_kmers.items():\n",
    "        left_occurrences = full_sequence[:start_index].count(kmer)\n",
    "        right_occurrences = full_sequence[end_index:].count(kmer)\n",
    "        occurrences = left_occurrences + right_occurrences\n",
    "        if occurrences == 0:\n",
    "            unique_kmers.append((kmer, details['count']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f0dcc7ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_unique_patterns(full_sequence, island_sequence, start_index, end_index, k_max):\n",
    "    \n",
    "    unique_kmers = []\n",
    "    \n",
    "    for k in range(4, k_max):\n",
    "        tfidf_matrix, vectorizer = tfidf(island_sequence, k)\n",
    "        kmers = get_kmers_map(tfidf_matrix, vectorizer, island_sequence)\n",
    "        find_unique_kmers(full_sequence, start_index, end_index, kmers, unique_kmers)\n",
    "        \n",
    "    return unique_kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8e9ecdb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "e_coli_data = read_data()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "89224e95",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Max pattern length: 100\n",
      "--------------------------\n",
      "Genome 1/90: NC_010468\n",
      "No PAI sequences\n",
      "CPAI 1/3: 1742 unique patterns\n",
      "CPAI 2/3: 1735 unique patterns\n",
      "CPAI 3/3: 631 unique patterns\n",
      "NPAI 1/2: 1743 unique patterns\n",
      "NPAI 2/2: 1746 unique patterns\n",
      "--------------------------\n",
      "Genome 2/90: NC_011748\n",
      "No PAI sequences\n",
      "CPAI 1/10: 641 unique patterns\n",
      "CPAI 2/10: 1760 unique patterns\n",
      "CPAI 3/10: 1751 unique patterns\n",
      "CPAI 4/10: 1749 unique patterns\n",
      "CPAI 5/10: 1619 unique patterns\n",
      "CPAI 6/10: 581 unique patterns\n",
      "CPAI 7/10: 36 unique patterns\n",
      "CPAI 8/10: 1638 unique patterns\n",
      "CPAI 9/10: 1648 unique patterns\n",
      "CPAI 10/10: 1066 unique patterns\n",
      "NPAI 1/1: 1740 unique patterns\n",
      "--------------------------\n",
      "Genome 3/90: NC_011751\n",
      "PAI 1/1: 1454 unique patterns\n",
      "CPAI 1/7: 657 unique patterns\n",
      "CPAI 2/7: 1754 unique patterns\n",
      "CPAI 3/7: 1738 unique patterns\n",
      "CPAI 4/7: 1745 unique patterns\n",
      "CPAI 5/7: 1753 unique patterns\n",
      "CPAI 6/7: 1748 unique patterns\n",
      "CPAI 7/7: 1399 unique patterns\n",
      "NPAI 1/2: 1748 unique patterns\n",
      "NPAI 2/2: 1737 unique patterns\n",
      "--------------------------\n",
      "Genome 4/90: NC_012759\n",
      "No PAI sequences\n",
      "CPAI 1/5: 1500 unique patterns\n",
      "CPAI 2/5: 1751 unique patterns\n",
      "CPAI 3/5: 1602 unique patterns\n",
      "CPAI 4/5: 1749 unique patterns\n",
      "CPAI 5/5: 1720 unique patterns\n",
      "NPAI 1/2: 1747 unique patterns\n",
      "NPAI 2/2: 1747 unique patterns\n",
      "--------------------------\n",
      "Genome 5/90: NC_017635\n",
      "No PAI sequences\n",
      "CPAI 1/4: 862 unique patterns\n",
      "CPAI 2/4: 1749 unique patterns\n",
      "CPAI 3/4: 1610 unique patterns\n",
      "CPAI 4/4: 1741 unique patterns\n",
      "NPAI 1/2: 1742 unique patterns\n",
      "NPAI 2/2: 1739 unique patterns\n",
      "--------------------------\n",
      "Genome 6/90: NC_013365\n",
      "No PAI sequences\n",
      "CPAI 1/1: 1801 unique patterns\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 7/90: NC_017641\n",
      "No PAI sequences\n",
      "CPAI 1/6: 666 unique patterns\n",
      "CPAI 2/6: 1728 unique patterns\n",
      "CPAI 3/6: 1761 unique patterns\n",
      "CPAI 4/6: 1746 unique patterns\n",
      "CPAI 5/6: 1733 unique patterns\n",
      "CPAI 6/6: 1685 unique patterns\n",
      "NPAI 1/1: 1721 unique patterns\n",
      "--------------------------\n",
      "Genome 8/90: NC_018662\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 1808 unique patterns\n",
      "--------------------------\n",
      "Genome 9/90: NC_010488\n",
      "No PAI sequences\n",
      "CPAI 1/1: 719 unique patterns\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 10/90: NC_017630\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 1809 unique patterns\n",
      "--------------------------\n",
      "Genome 11/90: NC_007946\n",
      "No PAI sequences\n",
      "CPAI 1/10: 666 unique patterns\n",
      "CPAI 2/10: 1743 unique patterns\n",
      "CPAI 3/10: 1752 unique patterns\n",
      "CPAI 4/10: 1742 unique patterns\n",
      "CPAI 5/10: 1749 unique patterns\n",
      "CPAI 6/10: 1730 unique patterns\n",
      "CPAI 7/10: 1747 unique patterns\n",
      "CPAI 8/10: 1750 unique patterns\n",
      "CPAI 9/10: 1754 unique patterns\n",
      "CPAI 10/10: 1748 unique patterns\n",
      "NPAI 1/2: 1746 unique patterns\n",
      "NPAI 2/2: 1748 unique patterns\n",
      "--------------------------\n",
      "Genome 12/90: NC_009786\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 13/90: NC_017660\n",
      "No PAI sequences\n",
      "CPAI 1/4: 1067 unique patterns\n",
      "CPAI 2/4: 1748 unique patterns\n",
      "CPAI 3/4: 1635 unique patterns\n",
      "CPAI 4/4: 1744 unique patterns\n",
      "NPAI 1/2: 1738 unique patterns\n",
      "NPAI 2/2: 1743 unique patterns\n",
      "--------------------------\n",
      "Genome 14/90: NC_017651\n",
      "No PAI sequences\n",
      "CPAI 1/12: 665 unique patterns\n",
      "CPAI 2/12: 1746 unique patterns\n",
      "CPAI 3/12: 1714 unique patterns\n",
      "CPAI 4/12: 1747 unique patterns\n",
      "CPAI 5/12: 1747 unique patterns\n",
      "CPAI 6/12: 1738 unique patterns\n",
      "CPAI 7/12: 1742 unique patterns\n",
      "CPAI 8/12: 1665 unique patterns\n",
      "CPAI 9/12: 1750 unique patterns\n",
      "CPAI 10/12: 1734 unique patterns\n",
      "CPAI 11/12: 1758 unique patterns\n",
      "CPAI 12/12: 1597 unique patterns\n",
      "NPAI 1/2: 1749 unique patterns\n",
      "NPAI 2/2: 1746 unique patterns\n",
      "--------------------------\n",
      "Genome 15/90: NC_002695\n",
      "PAI 1/2: 1743 unique patterns\n",
      "PAI 2/2: 1747 unique patterns\n",
      "CPAI 1/6: 682 unique patterns\n",
      "CPAI 2/6: 1351 unique patterns\n",
      "CPAI 3/6: 1749 unique patterns\n",
      "CPAI 4/6: 1743 unique patterns\n",
      "CPAI 5/6: 1489 unique patterns\n",
      "CPAI 6/6: 1747 unique patterns\n",
      "NPAI 1/2: 1696 unique patterns\n",
      "NPAI 2/2: 1740 unique patterns\n",
      "--------------------------\n",
      "Genome 16/90: NC_013353\n",
      "PAI 1/1: 972 unique patterns\n",
      "CPAI 1/8: 1745 unique patterns\n",
      "CPAI 2/8: 1741 unique patterns\n",
      "CPAI 3/8: 1710 unique patterns\n",
      "CPAI 4/8: 1751 unique patterns\n",
      "CPAI 5/8: 26 unique patterns\n",
      "CPAI 6/8: 1441 unique patterns\n",
      "CPAI 7/8: 1711 unique patterns\n",
      "CPAI 8/8: 1753 unique patterns\n",
      "NPAI 1/1: 1740 unique patterns\n",
      "--------------------------\n",
      "Genome 17/90: NC_007414\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 1804 unique patterns\n",
      "--------------------------\n",
      "Genome 18/90: NC_017656\n",
      "No PAI sequences\n",
      "CPAI 1/5: 665 unique patterns\n",
      "CPAI 2/5: 1411 unique patterns\n",
      "CPAI 3/5: 1750 unique patterns\n",
      "CPAI 4/5: 1629 unique patterns\n",
      "CPAI 5/5: 1749 unique patterns\n",
      "NPAI 1/2: 1751 unique patterns\n",
      "NPAI 2/2: 1741 unique patterns\n",
      "--------------------------\n",
      "Genome 19/90: NC_017664\n",
      "No PAI sequences\n",
      "CPAI 1/4: 862 unique patterns\n",
      "CPAI 2/4: 1753 unique patterns\n",
      "CPAI 3/4: 1610 unique patterns\n",
      "CPAI 4/4: 1741 unique patterns\n",
      "NPAI 1/2: 1740 unique patterns\n",
      "NPAI 2/2: 1739 unique patterns\n",
      "--------------------------\n",
      "Genome 20/90: NC_017659\n",
      "No PAI sequences\n",
      "CPAI 1/2: 1774 unique patterns\n",
      "CPAI 2/2: 1773 unique patterns\n",
      "NPAI 1/1: 1701 unique patterns\n",
      "--------------------------\n",
      "Genome 21/90: NC_000913\n",
      "No PAI sequences\n",
      "CPAI 1/4: 1527 unique patterns\n",
      "CPAI 2/4: 1324 unique patterns\n",
      "CPAI 3/4: 1720 unique patterns\n",
      "CPAI 4/4: 1752 unique patterns\n",
      "NPAI 1/3: 1602 unique patterns\n",
      "NPAI 2/3: 1746 unique patterns\n",
      "NPAI 3/3: 1748 unique patterns\n",
      "--------------------------\n",
      "Genome 22/90: NC_007941\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 1806 unique patterns\n",
      "--------------------------\n",
      "Genome 23/90: NC_012892\n",
      "No PAI sequences\n",
      "CPAI 1/6: 1741 unique patterns\n",
      "CPAI 2/6: 1763 unique patterns\n",
      "CPAI 3/6: 1750 unique patterns\n",
      "CPAI 4/6: 1742 unique patterns\n",
      "CPAI 5/6: 1757 unique patterns\n",
      "CPAI 6/6: 1488 unique patterns\n",
      "NPAI 1/1: 1688 unique patterns\n",
      "--------------------------\n",
      "Genome 24/90: NC_011601\n",
      "No PAI sequences\n",
      "CPAI 1/5: 1735 unique patterns\n",
      "CPAI 2/5: 1548 unique patterns\n",
      "CPAI 3/5: 1756 unique patterns\n",
      "CPAI 4/5: 1469 unique patterns\n",
      "CPAI 5/5: 1761 unique patterns\n",
      "NPAI 1/2: 1748 unique patterns\n",
      "NPAI 2/2: 1747 unique patterns\n",
      "--------------------------\n",
      "Genome 25/90: NC_011993\n",
      "No PAI sequences\n",
      "CPAI 1/6: 253 unique patterns\n",
      "CPAI 2/6: 1749 unique patterns\n",
      "CPAI 3/6: 1753 unique patterns\n",
      "CPAI 4/6: 1753 unique patterns\n",
      "CPAI 5/6: 1743 unique patterns\n",
      "CPAI 6/6: 1746 unique patterns\n",
      "NPAI 1/1: 1748 unique patterns\n",
      "--------------------------\n",
      "Genome 26/90: NC_011419\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 1798 unique patterns\n",
      "--------------------------\n",
      "Genome 27/90: NC_013364\n",
      "PAI 1/1: 1759 unique patterns\n",
      "CPAI 1/9: 955 unique patterns\n",
      "CPAI 2/9: 0 unique patterns\n",
      "CPAI 3/9: 17 unique patterns\n",
      "CPAI 4/9: 1625 unique patterns\n",
      "CPAI 5/9: 572 unique patterns\n",
      "CPAI 6/9: 1752 unique patterns\n",
      "CPAI 7/9: 1757 unique patterns\n",
      "CPAI 8/9: 1662 unique patterns\n",
      "CPAI 9/9: 1404 unique patterns\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 28/90: NC_007779\n",
      "No PAI sequences\n",
      "CPAI 1/4: 1521 unique patterns\n",
      "CPAI 2/4: 1751 unique patterns\n",
      "CPAI 3/4: 1720 unique patterns\n",
      "CPAI 4/4: 1586 unique patterns\n",
      "NPAI 1/3: 1600 unique patterns\n",
      "NPAI 2/3: 1744 unique patterns\n",
      "NPAI 3/3: 1750 unique patterns\n",
      "--------------------------\n",
      "Genome 29/90: NC_017645\n",
      "No PAI sequences\n",
      "CPAI 1/2: 1795 unique patterns\n",
      "CPAI 2/2: 1802 unique patterns\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 30/90: NC_013941\n",
      "No PAI sequences\n",
      "CPAI 1/5: 704 unique patterns\n",
      "CPAI 2/5: 1757 unique patterns\n",
      "CPAI 3/5: 1749 unique patterns\n",
      "CPAI 4/5: 1710 unique patterns\n",
      "CPAI 5/5: 1750 unique patterns\n",
      "NPAI 1/2: 1751 unique patterns\n",
      "NPAI 2/2: 1741 unique patterns\n",
      "--------------------------\n",
      "Genome 31/90: NC_009790\n",
      "No PAI sequences\n",
      "CPAI 1/2: 1616 unique patterns\n",
      "CPAI 2/2: 968 unique patterns\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 32/90: NC_012967\n",
      "No PAI sequences\n",
      "CPAI 1/7: 1741 unique patterns\n",
      "CPAI 2/7: 1763 unique patterns\n",
      "CPAI 3/7: 1683 unique patterns\n",
      "CPAI 4/7: 1742 unique patterns\n",
      "CPAI 5/7: 1756 unique patterns\n",
      "CPAI 6/7: 1487 unique patterns\n",
      "CPAI 7/7: 1753 unique patterns\n",
      "NPAI 1/1: 1687 unique patterns\n",
      "--------------------------\n",
      "Genome 33/90: NC_017628\n",
      "No PAI sequences\n",
      "CPAI 1/8: 656 unique patterns\n",
      "CPAI 2/8: 1745 unique patterns\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPAI 3/8: 1752 unique patterns\n",
      "CPAI 4/8: 1749 unique patterns\n",
      "CPAI 5/8: 1744 unique patterns\n",
      "CPAI 6/8: 1736 unique patterns\n",
      "CPAI 7/8: 1743 unique patterns\n",
      "CPAI 8/8: 1746 unique patterns\n",
      "NPAI 1/3: 1749 unique patterns\n",
      "NPAI 2/3: 1749 unique patterns\n",
      "NPAI 3/3: 1750 unique patterns\n",
      "--------------------------\n",
      "Genome 34/90: NC_009800\n",
      "No PAI sequences\n",
      "CPAI 1/4: 654 unique patterns\n",
      "CPAI 2/4: 1768 unique patterns\n",
      "CPAI 3/4: 1407 unique patterns\n",
      "CPAI 4/4: 1736 unique patterns\n",
      "NPAI 1/1: 1688 unique patterns\n",
      "--------------------------\n",
      "Genome 35/90: NC_004431\n",
      "PAI 1/2: 14 unique patterns\n",
      "PAI 2/2: 4 unique patterns\n",
      "CPAI 1/12: 664 unique patterns\n",
      "CPAI 2/12: 1716 unique patterns\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[5], line 30\u001b[0m\n\u001b[1;32m     28\u001b[0m genome_sequence \u001b[38;5;241m=\u001b[39m data[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgenome_sequence\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[1;32m     29\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m j, (cpai_sequence, indices) \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(data[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcpai_sequences\u001b[39m\u001b[38;5;124m'\u001b[39m]):\n\u001b[0;32m---> 30\u001b[0m     unique_patterns \u001b[38;5;241m=\u001b[39m \u001b[43mget_unique_patterns\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgenome_sequence\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcpai_sequence\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindices\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindices\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mk_max\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     32\u001b[0m     unique_cpai_patterns[name][\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcpai_sequence\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m cpai_sequence\n\u001b[1;32m     33\u001b[0m     unique_cpai_patterns[name][\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mindices\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m indices\n",
      "Cell \u001b[0;32mIn[3], line 8\u001b[0m, in \u001b[0;36mget_unique_patterns\u001b[0;34m(full_sequence, island_sequence, start_index, end_index, k_max)\u001b[0m\n\u001b[1;32m      6\u001b[0m     tfidf_matrix, vectorizer \u001b[38;5;241m=\u001b[39m tfidf(island_sequence, k)\n\u001b[1;32m      7\u001b[0m     kmers \u001b[38;5;241m=\u001b[39m get_kmers_map(tfidf_matrix, vectorizer, island_sequence)\n\u001b[0;32m----> 8\u001b[0m     \u001b[43mfind_unique_kmers\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfull_sequence\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstart_index\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mend_index\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkmers\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43munique_kmers\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     10\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m unique_kmers\n",
      "Cell \u001b[0;32mIn[2], line 32\u001b[0m, in \u001b[0;36mfind_unique_kmers\u001b[0;34m(full_sequence, start_index, end_index, significant_kmers, unique_kmers)\u001b[0m\n\u001b[1;32m     30\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m kmer, details \u001b[38;5;129;01min\u001b[39;00m significant_kmers\u001b[38;5;241m.\u001b[39mitems():\n\u001b[1;32m     31\u001b[0m     left_occurrences \u001b[38;5;241m=\u001b[39m full_sequence[:start_index]\u001b[38;5;241m.\u001b[39mcount(kmer)\n\u001b[0;32m---> 32\u001b[0m     right_occurrences \u001b[38;5;241m=\u001b[39m \u001b[43mfull_sequence\u001b[49m\u001b[43m[\u001b[49m\u001b[43mend_index\u001b[49m\u001b[43m:\u001b[49m\u001b[43m]\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcount\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkmer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     33\u001b[0m     occurrences \u001b[38;5;241m=\u001b[39m left_occurrences \u001b[38;5;241m+\u001b[39m right_occurrences\n\u001b[1;32m     34\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m occurrences \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m:\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/Bio/SeqRecord.py:981\u001b[0m, in \u001b[0;36mSeqRecord.count\u001b[0;34m(self, sub, start, end)\u001b[0m\n\u001b[1;32m    975\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mcount\u001b[39m(\u001b[38;5;28mself\u001b[39m, sub, start\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, end\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[1;32m    976\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Return the number of non-overlapping occurrences of sub in seq[start:end].\u001b[39;00m\n\u001b[1;32m    977\u001b[0m \n\u001b[1;32m    978\u001b[0m \u001b[38;5;124;03m    Optional arguments start and end are interpreted as in slice notation.\u001b[39;00m\n\u001b[1;32m    979\u001b[0m \u001b[38;5;124;03m    This method behaves as the count method of Python strings.\u001b[39;00m\n\u001b[1;32m    980\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 981\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mseq\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcount\u001b[49m\u001b[43m(\u001b[49m\u001b[43msub\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstart\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mend\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/Bio/Seq.py:648\u001b[0m, in \u001b[0;36m_SeqAbstractBaseClass.count\u001b[0;34m(self, sub, start, end)\u001b[0m\n\u001b[1;32m    643\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(sub, (\u001b[38;5;28mbytes\u001b[39m, \u001b[38;5;28mbytearray\u001b[39m)):\n\u001b[1;32m    644\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(\n\u001b[1;32m    645\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124ma Seq, MutableSeq, str, bytes, or bytearray object is required, not \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    646\u001b[0m         \u001b[38;5;241m%\u001b[39m \u001b[38;5;28mtype\u001b[39m(sub)\n\u001b[1;32m    647\u001b[0m     )\n\u001b[0;32m--> 648\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_data\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcount\u001b[49m\u001b[43m(\u001b[49m\u001b[43msub\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mstart\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mend\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "unique_pai_patterns  = {}\n",
    "unique_cpai_patterns = {}\n",
    "unique_npai_patterns = {}\n",
    "\n",
    "k_max = 100\n",
    "print(f'Max pattern length: {k_max}')\n",
    "print('--------------------------')\n",
    "\n",
    "for i, (name, data) in enumerate(e_coli_data.items()):\n",
    "    print(f'Genome {i+1}/{len(e_coli_data)}: {name}')\n",
    "    \n",
    "    if len(data['pai_sequences']) != 0:\n",
    "        unique_pai_patterns[name] = {}\n",
    "        genome_sequence = data['genome_sequence']\n",
    "        for j, (pai_sequence, indices) in enumerate(data['pai_sequences']):\n",
    "            unique_patterns = get_unique_patterns(genome_sequence, pai_sequence, indices[0], indices[1], k_max)\n",
    "\n",
    "            unique_pai_patterns[name]['pai_sequence'] = pai_sequence\n",
    "            unique_pai_patterns[name]['indices'] = indices\n",
    "            unique_pai_patterns[name]['patterns'] = unique_patterns\n",
    "\n",
    "            print(f\"PAI {j+1}/{len(data['pai_sequences'])}: {len(unique_patterns)} unique patterns\")\n",
    "    else:\n",
    "        print('No PAI sequences')\n",
    "    \n",
    "    if len(data['cpai_sequences']) != 0:    \n",
    "        unique_cpai_patterns[name] = {}\n",
    "        genome_sequence = data['genome_sequence']\n",
    "        for j, (cpai_sequence, indices) in enumerate(data['cpai_sequences']):\n",
    "            unique_patterns = get_unique_patterns(genome_sequence, cpai_sequence, indices[0], indices[1], k_max)\n",
    "\n",
    "            unique_cpai_patterns[name]['cpai_sequence'] = cpai_sequence\n",
    "            unique_cpai_patterns[name]['indices'] = indices\n",
    "            unique_cpai_patterns[name]['patterns'] = unique_patterns\n",
    "\n",
    "            print(f\"CPAI {j+1}/{len(data['cpai_sequences'])}: {len(unique_patterns)} unique patterns\")\n",
    "    else:\n",
    "        print('No CPAI sequences')\n",
    "        \n",
    "    if len(data['npai_sequences']) != 0:\n",
    "        unique_npai_patterns[name] = {}\n",
    "        genome_sequence = data['genome_sequence']\n",
    "        for j, (npai_sequence, indices) in enumerate(data['npai_sequences']):\n",
    "            unique_patterns = get_unique_patterns(genome_sequence, npai_sequence, indices[0], indices[1], k_max)\n",
    "\n",
    "            unique_npai_patterns[name]['npai_sequence'] = npai_sequence\n",
    "            unique_npai_patterns[name]['indices'] = indices\n",
    "            unique_npai_patterns[name]['patterns'] = unique_patterns\n",
    "\n",
    "            print(f\"NPAI {j+1}/{len(data['npai_sequences'])}: {len(unique_patterns)} unique patterns\")\n",
    "    else:\n",
    "        print('No NPAI sequences')\n",
    "    print('--------------------------')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e71bb949",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_pai_patterns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73e9ffcd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
