{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d358e28c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4029847e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_data():\n",
    "    base_directory = './h_pylori_paidb//'\n",
    "    data = {}\n",
    "    \n",
    "    for name in os.listdir(base_directory):\n",
    "        dir_path = os.path.join(base_directory, name)\n",
    "\n",
    "        if os.path.isdir(dir_path):\n",
    "            fasta_file = os.path.join(dir_path, f\"{name}.fasta\")\n",
    "            pai_file   = os.path.join(dir_path, f\"{name}.pai\")\n",
    "            cpai_file  = os.path.join(dir_path, f\"{name}.cpai\")\n",
    "            npai_file  = os.path.join(dir_path, f\"{name}.npai\")\n",
    "\n",
    "            data[name] = {}\n",
    "            genome_sequence = next(SeqIO.parse(fasta_file, 'fasta'))\n",
    "            data[name]['genome_sequence'] = genome_sequence\n",
    "\n",
    "            with open(pai_file, 'r') as file:\n",
    "                data[name]['pai_sequences'] = []\n",
    "\n",
    "                while True:\n",
    "                    line = file.readline()\n",
    "                    if not line:\n",
    "                        break\n",
    "\n",
    "                    line = line.split()\n",
    "                    pai_start_index = int(line[0])\n",
    "                    pai_end_index = int(line[1])\n",
    "\n",
    "                    pai_sequence = genome_sequence.seq[pai_start_index:pai_end_index]\n",
    "\n",
    "                    data[name]['pai_sequences'].append((pai_sequence, (pai_start_index, pai_end_index)))\n",
    "\n",
    "            with open(cpai_file, 'r') as file:\n",
    "                data[name]['cpai_sequences'] = []\n",
    "\n",
    "                while True:\n",
    "                    line = file.readline()\n",
    "                    if not line:\n",
    "                        break\n",
    "\n",
    "                    line = line.split()\n",
    "                    cpai_start_index = int(line[0])\n",
    "                    cpai_end_index = int(line[1])\n",
    "\n",
    "                    cpai_sequence = genome_sequence.seq[cpai_start_index:cpai_end_index]\n",
    "\n",
    "                    data[name]['cpai_sequences'].append((cpai_sequence, (cpai_start_index, cpai_end_index)))\n",
    "\n",
    "            with open(npai_file, 'r') as file:\n",
    "                data[name]['npai_sequences'] = []\n",
    "\n",
    "                while True:\n",
    "                    line = file.readline()\n",
    "                    if not line:\n",
    "                        break\n",
    "\n",
    "                    line = line.split()\n",
    "                    npai_start_index = int(line[0])\n",
    "                    npai_end_index = int(line[1])\n",
    "\n",
    "                    npai_sequence = genome_sequence.seq[npai_start_index:npai_end_index]\n",
    "\n",
    "                    data[name]['npai_sequences'].append((npai_sequence, (npai_start_index, npai_end_index)))\n",
    "\n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "48f67dc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "h_pylori_data = read_data()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea907e1d",
   "metadata": {},
   "source": [
    "### TF-IDF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3cb7ca41",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "\n",
    "def tfidf(sequence, k):\n",
    "    kmers = get_kmers(str(sequence), k)\n",
    "    vectorizer = TfidfVectorizer()\n",
    "    X = vectorizer.fit_transform([kmers])\n",
    "    tfidf_matrix = X.toarray()\n",
    "    return tfidf_matrix, vectorizer\n",
    "\n",
    "def get_kmers(sequence, k):\n",
    "    kmers = [sequence[i:i+k] for i in range(len(sequence) - k + 1)]\n",
    "    return ' '.join(kmers)\n",
    "\n",
    "def get_kmers_map(tfidf_matrix, vectorizer, sequence, top_n=20):\n",
    "    feature_names = vectorizer.get_feature_names_out()\n",
    "    feature_scores = tfidf_matrix.flatten()\n",
    "    \n",
    "    top_kmers_indices = feature_scores.argsort()[-top_n:][::-1]\n",
    "    top_kmers = [(feature_names[i], feature_scores[i]) for i in top_kmers_indices]\n",
    "\n",
    "    mapped_kmers = {}\n",
    "    for kmer, score in top_kmers:\n",
    "        kmer = kmer.upper()\n",
    "        count = sequence.count(kmer)\n",
    "        mapped_kmers[kmer] = {'score': score, 'count': count}\n",
    "        \n",
    "    return mapped_kmers\n",
    "\n",
    "def find_unique_kmers(full_sequence, start_index, end_index, significant_kmers, unique_kmers):\n",
    "    for kmer, details in significant_kmers.items():\n",
    "        left_occurrences = full_sequence[:start_index].count(kmer)\n",
    "        right_occurrences = full_sequence[end_index:].count(kmer)\n",
    "        occurrences = left_occurrences + right_occurrences\n",
    "        if occurrences == 0:\n",
    "            unique_kmers.append((kmer, details['count']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6db6afcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_unique_patterns(full_sequence, island_sequence, start_index, end_index, k_max):\n",
    "    \n",
    "    unique_kmers = []\n",
    "    \n",
    "    for k in range(4, k_max):\n",
    "        tfidf_matrix, vectorizer = tfidf(island_sequence, k)\n",
    "        kmers = get_kmers_map(tfidf_matrix, vectorizer, island_sequence)\n",
    "        find_unique_kmers(full_sequence, start_index, end_index, kmers, unique_kmers)\n",
    "        \n",
    "    return unique_kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6584609f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "\n",
    "logging.basicConfig(filename='./logs/tfidf_all_islands.log', level=logging.INFO,\n",
    "                    format='%(message)s')\n",
    "\n",
    "console = logging.StreamHandler()\n",
    "console.setLevel(logging.INFO)\n",
    "console.setFormatter(logging.Formatter('%(message)s'))\n",
    "\n",
    "logging.getLogger('').addHandler(console)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "195199f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Align\n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "432c2b84",
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_matches(sequence, pattern):\n",
    "#     aligner = Align.PairwiseAligner()\n",
    "#     aligner.mode = 'local'\n",
    "#     aligner.match_score = 2\n",
    "#     aligner.mismatch_score = -2\n",
    "#     aligner.open_gap_score = -2\n",
    "#     aligner.extend_gap_score = -1\n",
    "    \n",
    "#     alignments = aligner.align(sequence, pattern)\n",
    "#     threshold = 0.85 * (2 * len(pattern))\n",
    "    \n",
    "#     return sum(1 for alignment in alignments if alignment.score >= threshold)\n",
    "    return sequence.count(pattern)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4d476a11",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_islands_kmers(full_sequence,\n",
    "                       pai_islands,\n",
    "                       cpai_islands,\n",
    "                       npai_islands,\n",
    "                       significant_kmers,\n",
    "                       all_islands_kmers):\n",
    "    \n",
    "    all_islands = pai_islands + cpai_islands + npai_islands\n",
    "    all_islands.sort(key=lambda x: x[0])\n",
    "        \n",
    "    for kmer, details in significant_kmers.items():\n",
    "        \n",
    "        if kmer in all_islands_kmers:\n",
    "            continue\n",
    "        \n",
    "        start = 0\n",
    "        end = 0\n",
    "        isUnique = True\n",
    "        \n",
    "        # Check if kmer doesn't exist outside of the islands\n",
    "        for island in all_islands:\n",
    "            end = island[0]\n",
    "            \n",
    "            if end > start:\n",
    "                if count_matches(full_sequence[start:end], kmer) > 0:\n",
    "                    isUnique = False\n",
    "                    break\n",
    "                    \n",
    "            start = island[1]\n",
    "                    \n",
    "        if not isUnique or count_matches(full_sequence[start:], kmer) > 0:\n",
    "            continue\n",
    "          \n",
    "        \n",
    "        # Check which islands contain kmer\n",
    "        islands_containing_kmer = []\n",
    "        \n",
    "        for num, pai_island in enumerate(pai_islands, start=1):\n",
    "            start = pai_island[0]\n",
    "            end = pai_island[1]\n",
    "            \n",
    "            count = count_matches(full_sequence[start:end], kmer)\n",
    "            \n",
    "            if count > 0:\n",
    "                islands_containing_kmer.append(f'PAI {num}: {count}')\n",
    "            \n",
    "        for num, cpai_island in enumerate(cpai_islands, start=1):\n",
    "            start = cpai_island[0]\n",
    "            end = cpai_island[1]\n",
    "            \n",
    "            count = count_matches(full_sequence[start:end], kmer)\n",
    "            \n",
    "            if count > 0:\n",
    "                islands_containing_kmer.append(f'CPAI {num}: {count}')\n",
    "                \n",
    "        for num, npai_island in enumerate(npai_islands, start=1):\n",
    "            start = npai_island[0]\n",
    "            end = npai_island[1]\n",
    "            \n",
    "            count = count_matches(full_sequence[start:end], kmer)\n",
    "            \n",
    "            if count > 0:\n",
    "                islands_containing_kmer.append(f'NPAI {num}: {count}')\n",
    "        \n",
    "        all_islands_kmers[kmer] = islands_containing_kmer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "619814aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_islands_kmers(full_sequence,\n",
    "                      island_sequence,\n",
    "                      pai_islands,\n",
    "                      cpai_islands,\n",
    "                      npai_islands,\n",
    "                      all_islands_kmers,\n",
    "                      k_max):\n",
    "    \n",
    "    \n",
    "    for k in range(4, k_max):\n",
    "        tfidf_matrix, vectorizer = tfidf(island_sequence, k)\n",
    "        kmers = get_kmers_map(tfidf_matrix, vectorizer, island_sequence)\n",
    "        find_islands_kmers(full_sequence, pai_islands, cpai_islands, npai_islands, kmers, all_islands_kmers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "2bc4a3d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Max pattern length: 40\n",
      "--------------------------\n",
      "Genome 1/53: NC_017372\n",
      "No PAI sequences\n",
      "CPAI 1/1: 538 patterns found\n",
      "NPAI 1/1: 555 patterns found\n",
      "--------------------------\n",
      "Genome 2/53: NC_011333\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/2: 550 patterns found\n",
      "NPAI 2/2: 533 patterns found\n",
      "--------------------------\n",
      "Genome 3/53: NC_008086\n",
      "No PAI sequences\n",
      "CPAI 1/1: 542 patterns found\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 4/53: NC_021882\n",
      "No PAI sequences\n",
      "CPAI 1/1: 524 patterns found\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 5/53: NC_010698\n",
      "No PAI sequences\n",
      "CPAI 1/1: 524 patterns found\n",
      "NPAI 1/1: 550 patterns found\n",
      "--------------------------\n",
      "Genome 6/53: NC_017359\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 552 patterns found\n",
      "--------------------------\n",
      "Genome 7/53: NC_017355\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/2: 550 patterns found\n",
      "NPAI 2/2: 564 patterns found\n",
      "--------------------------\n",
      "Genome 8/53: NC_017375\n",
      "No PAI sequences\n",
      "CPAI 1/1: 557 patterns found\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 9/53: NC_017739\n",
      "No PAI sequences\n",
      "CPAI 1/1: 517 patterns found\n",
      "NPAI 1/2: 533 patterns found\n",
      "NPAI 2/2: 516 patterns found\n",
      "--------------------------\n",
      "Genome 10/53: NC_020508\n",
      "No PAI sequences\n",
      "CPAI 1/2: 531 patterns found\n",
      "CPAI 2/2: 522 patterns found\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 11/53: NC_017919\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 614 patterns found\n",
      "--------------------------\n",
      "Genome 12/53: NC_017382\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/2: 560 patterns found\n",
      "NPAI 2/2: 533 patterns found\n",
      "--------------------------\n",
      "Genome 13/53: NC_019563\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 537 patterns found\n",
      "--------------------------\n",
      "Genome 14/53: NC_017378\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 546 patterns found\n",
      "--------------------------\n",
      "Genome 15/53: NC_012973\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 462 patterns found\n",
      "--------------------------\n",
      "Genome 16/53: NC_021216\n",
      "No PAI sequences\n",
      "CPAI 1/1: 552 patterns found\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 17/53: NC_014560\n",
      "No PAI sequences\n",
      "CPAI 1/1: 526 patterns found\n",
      "NPAI 1/1: 554 patterns found\n",
      "--------------------------\n",
      "Genome 18/53: NC_021218\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/2: 522 patterns found\n",
      "NPAI 2/2: 545 patterns found\n",
      "--------------------------\n",
      "Genome 19/53: NC_017368\n",
      "PAI 1/1: 559 patterns found\n",
      "CPAI 1/1: 96 patterns found\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 20/53: NC_014256\n",
      "PAI 1/1: 559 patterns found\n",
      "CPAI 1/1: 533 patterns found\n",
      "NPAI 1/1: 103 patterns found\n",
      "--------------------------\n",
      "Genome 21/53: NC_020509\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 550 patterns found\n",
      "--------------------------\n",
      "Genome 22/53: NC_022130\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/2: 542 patterns found\n",
      "NPAI 2/2: 520 patterns found\n",
      "--------------------------\n",
      "Genome 23/53: NC_018939\n",
      "No PAI sequences\n",
      "CPAI 1/2: 546 patterns found\n",
      "CPAI 2/2: 552 patterns found\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 24/53: NC_017741\n",
      "No PAI sequences\n",
      "CPAI 1/3: 523 patterns found\n",
      "CPAI 2/3: 563 patterns found\n",
      "CPAI 3/3: 514 patterns found\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 25/53: NC_018937\n",
      "No PAI sequences\n",
      "CPAI 1/2: 546 patterns found\n",
      "CPAI 2/2: 552 patterns found\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 26/53: NC_000921\n",
      "PAI 1/1: 559 patterns found\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 78 patterns found\n",
      "--------------------------\n",
      "Genome 27/53: NC_017371\n",
      "No PAI sequences\n",
      "CPAI 1/1: 532 patterns found\n",
      "NPAI 1/1: 560 patterns found\n",
      "--------------------------\n",
      "Genome 28/53: NC_019560\n",
      "No PAI sequences\n",
      "CPAI 1/1: 267 patterns found\n",
      "NPAI 1/2: 545 patterns found\n",
      "NPAI 2/2: 541 patterns found\n",
      "--------------------------\n",
      "Genome 29/53: NC_017358\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/2: 529 patterns found\n",
      "NPAI 2/2: 549 patterns found\n",
      "--------------------------\n",
      "Genome 30/53: NC_021217\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 548 patterns found\n",
      "--------------------------\n",
      "Genome 31/53: NC_017733\n",
      "No PAI sequences\n",
      "CPAI 1/1: 546 patterns found\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 32/53: NC_017926\n",
      "No PAI sequences\n",
      "CPAI 1/2: 530 patterns found\n",
      "CPAI 2/2: 547 patterns found\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 33/53: NC_018938\n",
      "No PAI sequences\n",
      "CPAI 1/2: 546 patterns found\n",
      "CPAI 2/2: 552 patterns found\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 34/53: NC_017063\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 550 patterns found\n",
      "--------------------------\n",
      "Genome 35/53: NC_017366\n",
      "PAI 1/1: 540 patterns found\n",
      "CPAI 1/1: 76 patterns found\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 36/53: NC_017740\n",
      "No PAI sequences\n",
      "CPAI 1/1: 528 patterns found\n",
      "NPAI 1/1: 552 patterns found\n",
      "--------------------------\n",
      "Genome 37/53: NC_000915\n",
      "PAI 1/1: 551 patterns found\n",
      "CPAI 1/1: 545 patterns found\n",
      "NPAI 1/1: 110 patterns found\n",
      "--------------------------\n",
      "Genome 38/53: NC_017374\n",
      "No PAI sequences\n",
      "CPAI 1/1: 535 patterns found\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 39/53: NC_017376\n",
      "No PAI sequences\n",
      "CPAI 1/1: 526 patterns found\n",
      "NPAI 1/1: 549 patterns found\n",
      "--------------------------\n",
      "Genome 40/53: NC_017367\n",
      "PAI 1/1: 564 patterns found\n",
      "CPAI 1/2: 522 patterns found\n",
      "CPAI 2/2: 87 patterns found\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 41/53: NC_014555\n",
      "No PAI sequences\n",
      "CPAI 1/2: 559 patterns found\n",
      "CPAI 2/2: 522 patterns found\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 42/53: NC_017379\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/2: 540 patterns found\n",
      "NPAI 2/2: 540 patterns found\n",
      "--------------------------\n",
      "Genome 43/53: NC_017363\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 625 patterns found\n",
      "--------------------------\n",
      "Genome 44/53: NC_011498\n",
      "PAI 1/1: 517 patterns found\n",
      "No CPAI sequences\n",
      "NPAI 1/2: 0 patterns found\n",
      "NPAI 2/2: 549 patterns found\n",
      "--------------------------\n",
      "Genome 45/53: NC_017357\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 543 patterns found\n",
      "--------------------------\n",
      "Genome 46/53: NC_017742\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 545 patterns found\n",
      "--------------------------\n",
      "Genome 47/53: NC_017361\n",
      "No PAI sequences\n",
      "CPAI 1/2: 525 patterns found\n",
      "CPAI 2/2: 469 patterns found\n",
      "NPAI 1/1: 543 patterns found\n",
      "--------------------------\n",
      "Genome 48/53: NC_017381\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 537 patterns found\n",
      "--------------------------\n",
      "Genome 49/53: NC_017362\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 545 patterns found\n",
      "--------------------------\n",
      "Genome 50/53: NC_021215\n",
      "No PAI sequences\n",
      "CPAI 1/1: 446 patterns found\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 51/53: NC_017360\n",
      "No PAI sequences\n",
      "CPAI 1/1: 543 patterns found\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 52/53: NC_017365\n",
      "PAI 1/1: 557 patterns found\n",
      "CPAI 1/1: 105 patterns found\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 53/53: NC_017354\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 560 patterns found\n",
      "--------------------------\n"
     ]
    }
   ],
   "source": [
    "k_max = 40\n",
    "logging.info(f'Max pattern length: {k_max}')\n",
    "logging.info('--------------------------')\n",
    "\n",
    "unique_patterns = {}\n",
    "\n",
    "for i, (name, data) in enumerate(h_pylori_data.items()):\n",
    "    logging.info(f'Genome {i+1}/{len(h_pylori_data)}: {name}')\n",
    "    \n",
    "    genome_sequence = data['genome_sequence']\n",
    "    all_islands_kmers = {}\n",
    "\n",
    "    pai_indices = [x[1] for x in data['pai_sequences']]\n",
    "    cpai_indices = [x[1] for x in data['cpai_sequences']]\n",
    "    npai_indices = [x[1] for x in data['npai_sequences']]\n",
    "    \n",
    "    if len(data['pai_sequences']) != 0:\n",
    "        \n",
    "        for j, (pai_sequence, indices) in enumerate(data['pai_sequences']):\n",
    "            \n",
    "            len_before = len(all_islands_kmers)\n",
    "            \n",
    "            get_islands_kmers(genome_sequence,\n",
    "                              pai_sequence,\n",
    "                              pai_indices,\n",
    "                              cpai_indices,\n",
    "                              npai_indices,\n",
    "                              all_islands_kmers,\n",
    "                              k_max)\n",
    "            \n",
    "            new_patterns_found = len(all_islands_kmers) - len_before\n",
    "            logging.info(f\"PAI {j+1}/{len(data['pai_sequences'])}: {new_patterns_found} patterns found\")\n",
    "                        \n",
    "    else:\n",
    "        logging.info('No PAI sequences')\n",
    "    \n",
    "    \n",
    "    if len(data['cpai_sequences']) != 0:\n",
    "        \n",
    "        for j, (cpai_sequence, indices) in enumerate(data['cpai_sequences']):\n",
    "            \n",
    "            len_before = len(all_islands_kmers)\n",
    "            \n",
    "            get_islands_kmers(genome_sequence,\n",
    "                              cpai_sequence,\n",
    "                              pai_indices,\n",
    "                              cpai_indices,\n",
    "                              npai_indices,\n",
    "                              all_islands_kmers,\n",
    "                              k_max)\n",
    "            \n",
    "            new_patterns_found = len(all_islands_kmers) - len_before\n",
    "            logging.info(f\"CPAI {j+1}/{len(data['cpai_sequences'])}: {new_patterns_found} patterns found\")\n",
    "            \n",
    "    else:\n",
    "        logging.info('No CPAI sequences')\n",
    "        \n",
    "        \n",
    "    if len(data['npai_sequences']) != 0:\n",
    "\n",
    "        for j, (npai_sequence, indices) in enumerate(data['npai_sequences']):\n",
    "            \n",
    "            len_before = len(all_islands_kmers)\n",
    "            \n",
    "            get_islands_kmers(genome_sequence,\n",
    "                              npai_sequence,\n",
    "                              pai_indices,\n",
    "                              cpai_indices,\n",
    "                              npai_indices,\n",
    "                              all_islands_kmers,\n",
    "                              k_max)\n",
    "            \n",
    "            new_patterns_found = len(all_islands_kmers) - len_before\n",
    "            logging.info(f\"NPAI {j+1}/{len(data['npai_sequences'])}: {new_patterns_found} patterns found\")\n",
    "    else:\n",
    "        logging.info('No NPAI sequences')\n",
    "    logging.info('--------------------------')\n",
    "    \n",
    "    unique_patterns[name] = all_islands_kmers\n",
    "    os.makedirs(f'h_pylori_paidb/{name}/patterns/tfidf/all_islands/', exist_ok=True)\n",
    "    \n",
    "    with open(f'h_pylori_paidb/{name}/patterns/tfidf/all_islands/tfidf.txt', 'w') as file:\n",
    "        for kmer in all_islands_kmers:\n",
    "            file.write(kmer)\n",
    "            file.write('\\n')\n",
    "            \n",
    "            for island in all_islands_kmers[kmer]:\n",
    "                file.write(island)\n",
    "                file.write('\\n')\n",
    "            \n",
    "            file.write('----------------\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "94cf6c59",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3b1fa08f",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('./json/tfidf/all_islands.json', 'w') as file:\n",
    "    json.dump(unique_patterns, file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f199a33",
   "metadata": {},
   "source": [
    "#### Patterns that partially belong in the islands"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3f3505e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "\n",
    "logging.basicConfig(filename='./logs/tfidf_partial.log', level=logging.INFO,\n",
    "                    format='%(message)s')\n",
    "\n",
    "console = logging.StreamHandler()\n",
    "console.setLevel(logging.INFO)\n",
    "console.setFormatter(logging.Formatter('%(message)s'))\n",
    "\n",
    "logging.getLogger('').addHandler(console)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a474f524",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_unique_kmers_partial(full_sequence, island_start_index, island_end_index, significant_kmers, unique_kmers):\n",
    "    for kmer, details in significant_kmers.items():\n",
    "        \n",
    "        is_unique = True\n",
    "        is_partial = False\n",
    "        index = 0\n",
    "        \n",
    "        while index != -1:\n",
    "            index = full_sequence.seq.find(kmer, index)\n",
    "            if index != -1:\n",
    "                start = index\n",
    "                end = index + len(kmer)\n",
    "                \n",
    "                if start < island_start_index:\n",
    "                    if end < island_start_index:\n",
    "                        is_unique = False\n",
    "                        break\n",
    "                    else:\n",
    "                        is_partial = True\n",
    "                        \n",
    "                if end > island_end_index:\n",
    "                    if start > island_end_index:\n",
    "                        is_unique = False\n",
    "                        break\n",
    "                    else:\n",
    "                        is_partial = True\n",
    "                \n",
    "                index += len(kmer)\n",
    "                \n",
    "        if not is_unique or not is_partial:\n",
    "            continue\n",
    "        \n",
    "        left_occurrences = full_sequence[:island_start_index].count(kmer)\n",
    "        right_occurrences = full_sequence[island_end_index:].count(kmer)\n",
    "        occurrences = left_occurrences + right_occurrences\n",
    "        \n",
    "        if occurrences == 0:\n",
    "            unique_kmers.append((kmer, details['count']))\n",
    "        else:\n",
    "            print('Something went wrong')\n",
    "            \n",
    "def get_unique_patterns_partial(full_sequence, \n",
    "                                island_start_index,\n",
    "                                island_end_index,\n",
    "                                k_max):\n",
    "    \n",
    "    unique_kmers = []\n",
    "    \n",
    "    for k in range(4, k_max):\n",
    "        \n",
    "        start = island_start_index-k+1\n",
    "        end = island_end_index+k-1\n",
    "        \n",
    "        if start < 0:\n",
    "            start = 0\n",
    "            \n",
    "        if end > len(full_sequence.seq) - 1:\n",
    "            end = len(full_sequence.seq) - 1\n",
    "        \n",
    "        island_sequence = full_sequence.seq[start:end]\n",
    "        \n",
    "        tfidf_matrix, vectorizer = tfidf(island_sequence, k)\n",
    "        kmers = get_kmers_map(tfidf_matrix, vectorizer, island_sequence)\n",
    "        find_unique_kmers_partial(full_sequence, island_start_index, island_end_index, kmers, unique_kmers)\n",
    "        \n",
    "    return unique_kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ce1fd6c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Max pattern length: 40\n",
      "--------------------------\n",
      "Genome 1/53: NC_017372\n",
      "No PAI sequences\n",
      "CPAI 1/1: 0 unique patterns\n",
      "NPAI 1/1: 0 unique patterns\n",
      "--------------------------\n",
      "Genome 2/53: NC_011333\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/2: 0 unique patterns\n",
      "NPAI 2/2: 0 unique patterns\n",
      "--------------------------\n",
      "Genome 3/53: NC_008086\n",
      "No PAI sequences\n",
      "CPAI 1/1: 0 unique patterns\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 4/53: NC_021882\n",
      "No PAI sequences\n",
      "CPAI 1/1: 0 unique patterns\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 5/53: NC_010698\n",
      "No PAI sequences\n",
      "CPAI 1/1: 0 unique patterns\n",
      "NPAI 1/1: 0 unique patterns\n",
      "--------------------------\n",
      "Genome 6/53: NC_017359\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 0 unique patterns\n",
      "--------------------------\n",
      "Genome 7/53: NC_017355\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/2: 0 unique patterns\n",
      "NPAI 2/2: 2 unique patterns\n",
      "--------------------------\n",
      "Genome 8/53: NC_017375\n",
      "No PAI sequences\n",
      "CPAI 1/1: 0 unique patterns\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 9/53: NC_017739\n",
      "No PAI sequences\n",
      "CPAI 1/1: 0 unique patterns\n",
      "NPAI 1/2: 0 unique patterns\n",
      "NPAI 2/2: 0 unique patterns\n",
      "--------------------------\n",
      "Genome 10/53: NC_020508\n",
      "No PAI sequences\n",
      "CPAI 1/2: 0 unique patterns\n",
      "CPAI 2/2: 0 unique patterns\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 11/53: NC_017919\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 0 unique patterns\n",
      "--------------------------\n",
      "Genome 12/53: NC_017382\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/2: 0 unique patterns\n",
      "NPAI 2/2: 0 unique patterns\n",
      "--------------------------\n",
      "Genome 13/53: NC_019563\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 1 unique patterns\n",
      "--------------------------\n",
      "Genome 14/53: NC_017378\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 0 unique patterns\n",
      "--------------------------\n",
      "Genome 15/53: NC_012973\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 17 unique patterns\n",
      "--------------------------\n",
      "Genome 16/53: NC_021216\n",
      "No PAI sequences\n",
      "CPAI 1/1: 0 unique patterns\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 17/53: NC_014560\n",
      "No PAI sequences\n",
      "CPAI 1/1: 1 unique patterns\n",
      "NPAI 1/1: 0 unique patterns\n",
      "--------------------------\n",
      "Genome 18/53: NC_021218\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/2: 0 unique patterns\n",
      "NPAI 2/2: 0 unique patterns\n",
      "--------------------------\n",
      "Genome 19/53: NC_017368\n",
      "PAI 1/1: 0 unique patterns\n",
      "CPAI 1/1: 0 unique patterns\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 20/53: NC_014256\n",
      "PAI 1/1: 0 unique patterns\n",
      "CPAI 1/1: 0 unique patterns\n",
      "NPAI 1/1: 0 unique patterns\n",
      "--------------------------\n",
      "Genome 21/53: NC_020509\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 0 unique patterns\n",
      "--------------------------\n",
      "Genome 22/53: NC_022130\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/2: 0 unique patterns\n",
      "NPAI 2/2: 0 unique patterns\n",
      "--------------------------\n",
      "Genome 23/53: NC_018939\n",
      "No PAI sequences\n",
      "CPAI 1/2: 0 unique patterns\n",
      "CPAI 2/2: 0 unique patterns\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 24/53: NC_017741\n",
      "No PAI sequences\n",
      "CPAI 1/3: 22 unique patterns\n",
      "CPAI 2/3: 0 unique patterns\n",
      "CPAI 3/3: 0 unique patterns\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 25/53: NC_018937\n",
      "No PAI sequences\n",
      "CPAI 1/2: 0 unique patterns\n",
      "CPAI 2/2: 0 unique patterns\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 26/53: NC_000921\n",
      "PAI 1/1: 0 unique patterns\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 0 unique patterns\n",
      "--------------------------\n",
      "Genome 27/53: NC_017371\n",
      "No PAI sequences\n",
      "CPAI 1/1: 0 unique patterns\n",
      "NPAI 1/1: 0 unique patterns\n",
      "--------------------------\n",
      "Genome 28/53: NC_019560\n",
      "No PAI sequences\n",
      "CPAI 1/1: 0 unique patterns\n",
      "NPAI 1/2: 0 unique patterns\n",
      "NPAI 2/2: 0 unique patterns\n",
      "--------------------------\n",
      "Genome 29/53: NC_017358\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/2: 0 unique patterns\n",
      "NPAI 2/2: 0 unique patterns\n",
      "--------------------------\n",
      "Genome 30/53: NC_021217\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 0 unique patterns\n",
      "--------------------------\n",
      "Genome 31/53: NC_017733\n",
      "No PAI sequences\n",
      "CPAI 1/1: 0 unique patterns\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 32/53: NC_017926\n",
      "No PAI sequences\n",
      "CPAI 1/2: 0 unique patterns\n",
      "CPAI 2/2: 0 unique patterns\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 33/53: NC_018938\n",
      "No PAI sequences\n",
      "CPAI 1/2: 0 unique patterns\n",
      "CPAI 2/2: 0 unique patterns\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 34/53: NC_017063\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 0 unique patterns\n",
      "--------------------------\n",
      "Genome 35/53: NC_017366\n",
      "PAI 1/1: 0 unique patterns\n",
      "CPAI 1/1: 0 unique patterns\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 36/53: NC_017740\n",
      "No PAI sequences\n",
      "CPAI 1/1: 0 unique patterns\n",
      "NPAI 1/1: 0 unique patterns\n",
      "--------------------------\n",
      "Genome 37/53: NC_000915\n",
      "PAI 1/1: 0 unique patterns\n",
      "CPAI 1/1: 0 unique patterns\n",
      "NPAI 1/1: 0 unique patterns\n",
      "--------------------------\n",
      "Genome 38/53: NC_017374\n",
      "No PAI sequences\n",
      "CPAI 1/1: 0 unique patterns\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 39/53: NC_017376\n",
      "No PAI sequences\n",
      "CPAI 1/1: 0 unique patterns\n",
      "NPAI 1/1: 0 unique patterns\n",
      "--------------------------\n",
      "Genome 40/53: NC_017367\n",
      "PAI 1/1: 0 unique patterns\n",
      "CPAI 1/2: 0 unique patterns\n",
      "CPAI 2/2: 0 unique patterns\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 41/53: NC_014555\n",
      "No PAI sequences\n",
      "CPAI 1/2: 0 unique patterns\n",
      "CPAI 2/2: 0 unique patterns\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 42/53: NC_017379\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/2: 0 unique patterns\n",
      "NPAI 2/2: 0 unique patterns\n",
      "--------------------------\n",
      "Genome 43/53: NC_017363\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 0 unique patterns\n",
      "--------------------------\n",
      "Genome 44/53: NC_011498\n",
      "PAI 1/1: 0 unique patterns\n",
      "No CPAI sequences\n",
      "NPAI 1/2: 0 unique patterns\n",
      "NPAI 2/2: 0 unique patterns\n",
      "--------------------------\n",
      "Genome 45/53: NC_017357\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 0 unique patterns\n",
      "--------------------------\n",
      "Genome 46/53: NC_017742\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 0 unique patterns\n",
      "--------------------------\n",
      "Genome 47/53: NC_017361\n",
      "No PAI sequences\n",
      "CPAI 1/2: 0 unique patterns\n",
      "CPAI 2/2: 0 unique patterns\n",
      "NPAI 1/1: 0 unique patterns\n",
      "--------------------------\n",
      "Genome 48/53: NC_017381\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 0 unique patterns\n",
      "--------------------------\n",
      "Genome 49/53: NC_017362\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 0 unique patterns\n",
      "--------------------------\n",
      "Genome 50/53: NC_021215\n",
      "No PAI sequences\n",
      "CPAI 1/1: 0 unique patterns\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 51/53: NC_017360\n",
      "No PAI sequences\n",
      "CPAI 1/1: 0 unique patterns\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 52/53: NC_017365\n",
      "PAI 1/1: 0 unique patterns\n",
      "CPAI 1/1: 0 unique patterns\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 53/53: NC_017354\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: 0 unique patterns\n",
      "--------------------------\n"
     ]
    }
   ],
   "source": [
    "unique_pai_patterns  = {}\n",
    "unique_cpai_patterns = {}\n",
    "unique_npai_patterns = {}\n",
    "\n",
    "k_max = 40\n",
    "logging.info(f'Max pattern length: {k_max}')\n",
    "logging.info('--------------------------')\n",
    "\n",
    "for i, (name, data) in enumerate(h_pylori_data.items()):        \n",
    "    logging.info(f'Genome {i+1}/{len(h_pylori_data)}: {name}')\n",
    "    \n",
    "    if len(data['pai_sequences']) != 0:\n",
    "        unique_pai_patterns[name] = {}\n",
    "        genome_sequence = data['genome_sequence']\n",
    "        for j, (pai_sequence, indices) in enumerate(data['pai_sequences']):\n",
    "            unique_patterns = get_unique_patterns_partial(genome_sequence, indices[0], indices[1], k_max)\n",
    "\n",
    "            unique_pai_patterns[name][f'pai_sequence_{j+1}'] = pai_sequence\n",
    "            unique_pai_patterns[name][f'indices_pai_{j+1}'] = indices\n",
    "            unique_pai_patterns[name][f'patterns_pai_{j+1}'] = unique_patterns\n",
    "\n",
    "            logging.info(f\"PAI {j+1}/{len(data['pai_sequences'])}: {len(unique_patterns)} unique patterns\")\n",
    "            \n",
    "            os.makedirs(f'h_pylori_paidb/{name}/patterns/tfidf/partial/pai/', exist_ok=True)\n",
    "            \n",
    "            with open(f'h_pylori_paidb/{name}/patterns/tfidf/partial/pai/pai_{j+1}_patterns.txt', 'w') as file:\n",
    "                for unique_pattern, frequency in unique_patterns:\n",
    "                    file.write(f'{unique_pattern}: {frequency}\\n')\n",
    "                        \n",
    "    else:\n",
    "        logging.info('No PAI sequences')\n",
    "    \n",
    "    if len(data['cpai_sequences']) != 0:    \n",
    "        unique_cpai_patterns[name] = {}\n",
    "\n",
    "        genome_sequence = data['genome_sequence']\n",
    "        for j, (cpai_sequence, indices) in enumerate(data['cpai_sequences']):\n",
    "            unique_patterns = get_unique_patterns_partial(genome_sequence, indices[0], indices[1], k_max)\n",
    "\n",
    "            unique_cpai_patterns[name][f'cpai_sequence_{j+1}'] = cpai_sequence\n",
    "            unique_cpai_patterns[name][f'indices_cpai_{j+1}'] = indices\n",
    "            unique_cpai_patterns[name][f'patterns_cpai_{j+1}'] = unique_patterns\n",
    "            \n",
    "            logging.info(f\"CPAI {j+1}/{len(data['cpai_sequences'])}: {len(unique_patterns)} unique patterns\")\n",
    "            \n",
    "            os.makedirs(f'h_pylori_paidb/{name}/patterns/tfidf/partial/cpai/', exist_ok=True)\n",
    "            \n",
    "            with open(f'h_pylori_paidb/{name}/patterns/tfidf/partial/cpai/cpai_{j+1}_patterns.txt', 'w') as file:\n",
    "                for unique_pattern, frequency in unique_patterns:\n",
    "                    file.write(f'{unique_pattern}: {frequency}\\n')\n",
    "            \n",
    "    else:\n",
    "        logging.info('No CPAI sequences')\n",
    "        \n",
    "    if len(data['npai_sequences']) != 0:\n",
    "        unique_npai_patterns[name] = {}\n",
    "        genome_sequence = data['genome_sequence']\n",
    "        for j, (npai_sequence, indices) in enumerate(data['npai_sequences']):\n",
    "            unique_patterns = get_unique_patterns_partial(genome_sequence, indices[0], indices[1], k_max)\n",
    "\n",
    "            unique_npai_patterns[name][f'npai_sequence_{j+1}'] = npai_sequence\n",
    "            unique_npai_patterns[name][f'indices_npai_{j+1}'] = indices\n",
    "            unique_npai_patterns[name][f'patterns_npai_{j+1}'] = unique_patterns\n",
    "\n",
    "            logging.info(f\"NPAI {j+1}/{len(data['npai_sequences'])}: {len(unique_patterns)} unique patterns\")\n",
    "            \n",
    "            os.makedirs(f'h_pylori_paidb/{name}/patterns/tfidf/partial/npai/', exist_ok=True)\n",
    "            \n",
    "            with open(f'h_pylori_paidb/{name}/patterns/tfidf/partial/npai/npai_{j+1}_patterns.txt', 'w') as file:\n",
    "                for unique_pattern, frequency in unique_patterns:\n",
    "                    file.write(f'{unique_pattern}: {frequency}\\n')\n",
    "    \n",
    "    else:\n",
    "        logging.info('No NPAI sequences')\n",
    "    logging.info('--------------------------')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea2985f2",
   "metadata": {},
   "source": [
    "### NOSEP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "44683b12",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spmf import Spmf\n",
    "import os\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a1e79669",
   "metadata": {},
   "outputs": [],
   "source": [
    "def to_spmf_format(genome_sequence):\n",
    "    seq = str(genome_sequence)\n",
    "    \n",
    "    seq = seq.replace('A', '1')\n",
    "    seq = seq.replace('C', '2')\n",
    "    seq = seq.replace('G', '3')\n",
    "    seq = seq.replace('T', '4')\n",
    "\n",
    "    seq = ''.join(c + ' -1 ' for c in seq).rstrip(' -1 ') + ' -2'\n",
    "    \n",
    "    return seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b7259176",
   "metadata": {},
   "outputs": [],
   "source": [
    "def from_spmf_format(filename):\n",
    "    patterns = []\n",
    "    \n",
    "    with open(filename, 'r') as file:\n",
    "        for line in file:\n",
    "            lst = line.split('#')\n",
    "            \n",
    "            pattern = lst[0]\n",
    "            support = lst[1]\n",
    "            \n",
    "            pattern = pattern.replace('-1', '')\n",
    "            pattern = pattern.replace('1', 'A')\n",
    "            pattern = pattern.replace('2', 'C')\n",
    "            pattern = pattern.replace('3', 'G')\n",
    "            pattern = pattern.replace('4', 'T')\n",
    "            pattern = pattern.replace(' ', '')\n",
    "            \n",
    "            support = int(support.replace('SUP: ', ''))\n",
    "            patterns.append((pattern, support))\n",
    "            \n",
    "    return patterns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "81fa2e28",
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "\n",
    "logging.basicConfig(filename='logs/nosep.log', level=logging.INFO,\n",
    "                    format='%(message)s')\n",
    "\n",
    "console = logging.StreamHandler()\n",
    "console.setLevel(logging.INFO)\n",
    "console.setFormatter(logging.Formatter('%(message)s'))\n",
    "\n",
    "logging.getLogger('').addHandler(console)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1f548023",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_nosep(full_sequence, island_sequence, start_index, end_index, k_max):\n",
    "    \n",
    "    unique_patterns = []\n",
    "    \n",
    "    input_filename = 'input.txt'\n",
    "    output_filename = 'output.txt'\n",
    "    \n",
    "    with open(input_filename, 'w') as input_file:\n",
    "        input_file.write(to_spmf_format(island_sequence))\n",
    "    \n",
    "    spmf = Spmf(\"NOSEP\", input_filename=input_filename, output_filename=output_filename, arguments=[4, k_max, 0, 0, 2])\n",
    "    spmf.run()\n",
    "    \n",
    "    patterns = from_spmf_format(output_filename)\n",
    "    \n",
    "    os.remove(input_filename)\n",
    "    os.remove(output_filename)\n",
    "    \n",
    "    for pattern, support in patterns:\n",
    "        left_occurrences = full_sequence[:start_index].count(pattern)\n",
    "        right_occurrences = full_sequence[end_index:].count(pattern)\n",
    "        occurrences = left_occurrences + right_occurrences\n",
    "        if occurrences == 0:\n",
    "            unique_patterns.append((pattern, support))\n",
    "            \n",
    "    return unique_patterns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a85aa029",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Max pattern length: 40\n",
      "--------------------------\n",
      "Genome 1/53: NC_017372\n",
      "No PAI sequences\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 13971\n",
      " Total time ~ 34880 ms\n",
      " Maximum memory usage : 294.0252990722656 mb\n",
      " Calculation count 27737\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CPAI 1/1: 314 unique patterns\n",
      "NPAI 1/1: island too large\n",
      "--------------------------\n",
      "Genome 2/53: NC_011333\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/2: island too large\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 29246\n",
      " Total time ~ 176465 ms\n",
      " Maximum memory usage : 42.65605163574219 mb\n",
      " Calculation count 58527\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "NPAI 2/2: 594 unique patterns\n",
      "--------------------------\n",
      "Genome 3/53: NC_008086\n",
      "No PAI sequences\n",
      "CPAI 1/1: island too large\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 4/53: NC_021882\n",
      "No PAI sequences\n",
      "CPAI 1/1: island too large\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 5/53: NC_010698\n",
      "No PAI sequences\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 28954\n",
      " Total time ~ 189008 ms\n",
      " Maximum memory usage : 42.34050750732422 mb\n",
      " Calculation count 58091\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CPAI 1/1: 568 unique patterns\n",
      "NPAI 1/1: island too large\n",
      "--------------------------\n",
      "Genome 6/53: NC_017359\n",
      "No PAI sequences\n",
      "No CPAI sequences\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 46308\n",
      " Total time ~ 252636 ms\n",
      " Maximum memory usage : 391.41576385498047 mb\n",
      " Calculation count 74590\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "NPAI 1/1: 18858 unique patterns\n",
      "--------------------------\n",
      "Genome 7/53: NC_017355\n",
      "No PAI sequences\n",
      "No CPAI sequences\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 39876\n",
      " Total time ~ 167029 ms\n",
      " Maximum memory usage : 326.35790252685547 mb\n",
      " Calculation count 61935\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "NPAI 1/2: 18216 unique patterns\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 106609\n",
      " Total time ~ 378097 ms\n",
      " Maximum memory usage : 292.7956771850586 mb\n",
      " Calculation count 124299\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "NPAI 2/2: 81965 unique patterns\n",
      "--------------------------\n",
      "Genome 8/53: NC_017375\n",
      "No PAI sequences\n",
      "CPAI 1/1: island too large\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 9/53: NC_017739\n",
      "No PAI sequences\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 13198\n",
      " Total time ~ 39231 ms\n",
      " Maximum memory usage : 237.73016357421875 mb\n",
      " Calculation count 26213\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CPAI 1/1: 33 unique patterns\n",
      "NPAI 1/2: island too large\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 29056\n",
      " Total time ~ 250640 ms\n",
      " Maximum memory usage : 131.96405792236328 mb\n",
      " Calculation count 58279\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "NPAI 2/2: 406 unique patterns\n",
      "--------------------------\n",
      "Genome 10/53: NC_020508\n",
      "No PAI sequences\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 12799\n",
      " Total time ~ 35149 ms\n",
      " Maximum memory usage : 213.4578628540039 mb\n",
      " Calculation count 25661\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CPAI 1/2: 176 unique patterns\n",
      "CPAI 2/2: island too large\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 11/53: NC_017919\n",
      "No PAI sequences\n",
      "No CPAI sequences\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 15120\n",
      " Total time ~ 47002 ms\n",
      " Maximum memory usage : 166.27306365966797 mb\n",
      " Calculation count 30281\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "NPAI 1/1: 10269 unique patterns\n",
      "--------------------------\n",
      "Genome 12/53: NC_017382\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/2: island too large\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 14192\n",
      " Total time ~ 45104 ms\n",
      " Maximum memory usage : 120.14141845703125 mb\n",
      " Calculation count 28314\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "NPAI 2/2: 172 unique patterns\n",
      "--------------------------\n",
      "Genome 13/53: NC_019563\n",
      "No PAI sequences\n",
      "No CPAI sequences\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 7285\n",
      " Total time ~ 11327 ms\n",
      " Maximum memory usage : 60.97782897949219 mb\n",
      " Calculation count 14973\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "NPAI 1/1: 58 unique patterns\n",
      "--------------------------\n",
      "Genome 14/53: NC_017378\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: island too large\n",
      "--------------------------\n",
      "Genome 15/53: NC_012973\n",
      "No PAI sequences\n",
      "No CPAI sequences\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 9340\n",
      " Total time ~ 17689 ms\n",
      " Maximum memory usage : 184.47708892822266 mb\n",
      " Calculation count 19272\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "NPAI 1/1: 115 unique patterns\n",
      "--------------------------\n",
      "Genome 16/53: NC_021216\n",
      "No PAI sequences\n",
      "CPAI 1/1: island too large\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 17/53: NC_014560\n",
      "No PAI sequences\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 14651\n",
      " Total time ~ 45997 ms\n",
      " Maximum memory usage : 265.0581588745117 mb\n",
      " Calculation count 29401\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CPAI 1/1: 144 unique patterns\n",
      "NPAI 1/1: island too large\n",
      "--------------------------\n",
      "Genome 18/53: NC_021218\n",
      "No PAI sequences\n",
      "No CPAI sequences\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 28853\n",
      " Total time ~ 199299 ms\n",
      " Maximum memory usage : 39.043426513671875 mb\n",
      " Calculation count 58065\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "NPAI 1/2: 548 unique patterns\n",
      "NPAI 2/2: island too large\n",
      "--------------------------\n",
      "Genome 19/53: NC_017368\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 51506\n",
      " Total time ~ 296749 ms\n",
      " Maximum memory usage : 285.5534133911133 mb\n",
      " Calculation count 81123\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "PAI 1/1: 22999 unique patterns\n",
      "CPAI 1/1: island too large\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 20/53: NC_014256\n",
      "PAI 1/1: island too large\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 17635\n",
      " Total time ~ 60832 ms\n",
      " Maximum memory usage : 78.42208099365234 mb\n",
      " Calculation count 35360\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CPAI 1/1: 189 unique patterns\n",
      "NPAI 1/1: island too large\n",
      "--------------------------\n",
      "Genome 21/53: NC_020509\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: island too large\n",
      "--------------------------\n",
      "Genome 22/53: NC_022130\n",
      "No PAI sequences\n",
      "No CPAI sequences\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 14929\n",
      " Total time ~ 33492 ms\n",
      " Maximum memory usage : 89.73841857910156 mb\n",
      " Calculation count 27178\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "NPAI 1/2: 3097 unique patterns\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 29229\n",
      " Total time ~ 178529 ms\n",
      " Maximum memory usage : 368.30967712402344 mb\n",
      " Calculation count 58498\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "NPAI 2/2: 708 unique patterns\n",
      "--------------------------\n",
      "Genome 23/53: NC_018939\n",
      "No PAI sequences\n",
      "java.lang.NumberFormatException: For input string: \"Y\"\n",
      "\tat java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:67)\n",
      "\tat java.base/java.lang.Integer.parseInt(Integer.java:662)\n",
      "\tat java.base/java.lang.Integer.parseInt(Integer.java:778)\n",
      "\tat ca.pfv.spmf.algorithms.sequentialpatterns.nosep.AlgoNOSEP.readInputFile(AlgoNOSEP.java:809)\n",
      "\tat ca.pfv.spmf.algorithms.sequentialpatterns.nosep.AlgoNOSEP.runAlgorithm(AlgoNOSEP.java:878)\n",
      "\tat ca.pfv.spmf.algorithmmanager.descriptions.DescriptionAlgoNOSEP.runAlgorithm(DescriptionAlgoNOSEP.java:52)\n",
      "\tat ca.pfv.spmf.gui.CommandProcessor.runAlgorithm(CommandProcessor.java:265)\n",
      "\tat ca.pfv.spmf.gui.Main.processCommandLineArguments(Main.java:151)\n",
      "\tat ca.pfv.spmf.gui.Main.main(Main.java:64)\n",
      "CPAI 1/2: 0 unique patterns\n",
      "CPAI 2/2: island too large\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 24/53: NC_017741\n",
      "No PAI sequences\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 0\n",
      " Total time ~ 9 ms\n",
      " Maximum memory usage : 10.952568054199219 mb\n",
      " Calculation count 0\n",
      "===================================================\n",
      "\n",
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 13572\n",
      " Total time ~ 34770 ms\n",
      " Maximum memory usage : 33.820091247558594 mb\n",
      " Calculation count 26862\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CPAI 1/3: 51 unique patterns\n",
      "CPAI 2/3: island too large\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 28584\n",
      " Total time ~ 169700 ms\n",
      " Maximum memory usage : 176.6141128540039 mb\n",
      " Calculation count 57265\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CPAI 3/3: 367 unique patterns\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 25/53: NC_018937\n",
      "No PAI sequences\n",
      "java.lang.NumberFormatException: For input string: \"Y\"\n",
      "\tat java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:67)\n",
      "\tat java.base/java.lang.Integer.parseInt(Integer.java:662)\n",
      "\tat java.base/java.lang.Integer.parseInt(Integer.java:778)\n",
      "\tat ca.pfv.spmf.algorithms.sequentialpatterns.nosep.AlgoNOSEP.readInputFile(AlgoNOSEP.java:809)\n",
      "\tat ca.pfv.spmf.algorithms.sequentialpatterns.nosep.AlgoNOSEP.runAlgorithm(AlgoNOSEP.java:878)\n",
      "\tat ca.pfv.spmf.algorithmmanager.descriptions.DescriptionAlgoNOSEP.runAlgorithm(DescriptionAlgoNOSEP.java:52)\n",
      "\tat ca.pfv.spmf.gui.CommandProcessor.runAlgorithm(CommandProcessor.java:265)\n",
      "\tat ca.pfv.spmf.gui.Main.processCommandLineArguments(Main.java:151)\n",
      "\tat ca.pfv.spmf.gui.Main.main(Main.java:64)\n",
      "CPAI 1/2: 0 unique patterns\n",
      "CPAI 2/2: island too large\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 26/53: NC_000921\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 0\n",
      " Total time ~ 9 ms\n",
      " Maximum memory usage : 10.952568054199219 mb\n",
      " Calculation count 0\n",
      "===================================================\n",
      "\n",
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 47737\n",
      " Total time ~ 13352082 ms\n",
      " Maximum memory usage : 196.75147247314453 mb\n",
      " Calculation count 78630\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "PAI 1/1: 17978 unique patterns\n",
      "No CPAI sequences\n",
      "NPAI 1/1: island too large\n",
      "--------------------------\n",
      "Genome 27/53: NC_017371\n",
      "No PAI sequences\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 25457\n",
      " Total time ~ 129126 ms\n",
      " Maximum memory usage : 348.8876419067383 mb\n",
      " Calculation count 51172\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CPAI 1/1: 707 unique patterns\n",
      "NPAI 1/1: island too large\n",
      "--------------------------\n",
      "Genome 28/53: NC_019560\n",
      "No PAI sequences\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 12561\n",
      " Total time ~ 30257 ms\n",
      " Maximum memory usage : 199.0332489013672 mb\n",
      " Calculation count 24973\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CPAI 1/1: 45 unique patterns\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 6553\n",
      " Total time ~ 8153 ms\n",
      " Maximum memory usage : 40.442726135253906 mb\n",
      " Calculation count 13485\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "NPAI 1/2: 34 unique patterns\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 14595\n",
      " Total time ~ 40415 ms\n",
      " Maximum memory usage : 268.6669464111328 mb\n",
      " Calculation count 29057\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "NPAI 2/2: 171 unique patterns\n",
      "--------------------------\n",
      "Genome 29/53: NC_017358\n",
      "No PAI sequences\n",
      "No CPAI sequences\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 28591\n",
      " Total time ~ 167196 ms\n",
      " Maximum memory usage : 247.0511703491211 mb\n",
      " Calculation count 57280\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "NPAI 1/2: 537 unique patterns\n",
      "NPAI 2/2: island too large\n",
      "--------------------------\n",
      "Genome 30/53: NC_021217\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: island too large\n",
      "--------------------------\n",
      "Genome 31/53: NC_017733\n",
      "No PAI sequences\n",
      "CPAI 1/1: island too large\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 32/53: NC_017926\n",
      "No PAI sequences\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 23689\n",
      " Total time ~ 276966 ms\n",
      " Maximum memory usage : 240.13111114501953 mb\n",
      " Calculation count 47607\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CPAI 1/2: 278 unique patterns\n",
      "CPAI 2/2: island too large\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 33/53: NC_018938\n",
      "No PAI sequences\n",
      "java.lang.NumberFormatException: For input string: \"Y\"\n",
      "\tat java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:67)\n",
      "\tat java.base/java.lang.Integer.parseInt(Integer.java:662)\n",
      "\tat java.base/java.lang.Integer.parseInt(Integer.java:778)\n",
      "\tat ca.pfv.spmf.algorithms.sequentialpatterns.nosep.AlgoNOSEP.readInputFile(AlgoNOSEP.java:809)\n",
      "\tat ca.pfv.spmf.algorithms.sequentialpatterns.nosep.AlgoNOSEP.runAlgorithm(AlgoNOSEP.java:878)\n",
      "\tat ca.pfv.spmf.algorithmmanager.descriptions.DescriptionAlgoNOSEP.runAlgorithm(DescriptionAlgoNOSEP.java:52)\n",
      "\tat ca.pfv.spmf.gui.CommandProcessor.runAlgorithm(CommandProcessor.java:265)\n",
      "\tat ca.pfv.spmf.gui.Main.processCommandLineArguments(Main.java:151)\n",
      "\tat ca.pfv.spmf.gui.Main.main(Main.java:64)\n",
      "CPAI 1/2: 0 unique patterns\n",
      "CPAI 2/2: island too large\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 34/53: NC_017063\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: island too large\n",
      "--------------------------\n",
      "Genome 35/53: NC_017366\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 0\n",
      " Total time ~ 33 ms\n",
      " Maximum memory usage : 10.952568054199219 mb\n",
      " Calculation count 0\n",
      "===================================================\n",
      "\n",
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 58318\n",
      " Total time ~ 958458 ms\n",
      " Maximum memory usage : 175.55028533935547 mb\n",
      " Calculation count 87687\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "PAI 1/1: 29601 unique patterns\n",
      "CPAI 1/1: island too large\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 36/53: NC_017740\n",
      "No PAI sequences\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 28993\n",
      " Total time ~ 173417 ms\n",
      " Maximum memory usage : 243.3825225830078 mb\n",
      " Calculation count 58136\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CPAI 1/1: 574 unique patterns\n",
      "NPAI 1/1: island too large\n",
      "--------------------------\n",
      "Genome 37/53: NC_000915\n",
      "java.lang.NumberFormatException: For input string: \"K\"\n",
      "\tat java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:67)\n",
      "\tat java.base/java.lang.Integer.parseInt(Integer.java:662)\n",
      "\tat java.base/java.lang.Integer.parseInt(Integer.java:778)\n",
      "\tat ca.pfv.spmf.algorithms.sequentialpatterns.nosep.AlgoNOSEP.readInputFile(AlgoNOSEP.java:809)\n",
      "\tat ca.pfv.spmf.algorithms.sequentialpatterns.nosep.AlgoNOSEP.runAlgorithm(AlgoNOSEP.java:878)\n",
      "\tat ca.pfv.spmf.algorithmmanager.descriptions.DescriptionAlgoNOSEP.runAlgorithm(DescriptionAlgoNOSEP.java:52)\n",
      "\tat ca.pfv.spmf.gui.CommandProcessor.runAlgorithm(CommandProcessor.java:265)\n",
      "\tat ca.pfv.spmf.gui.Main.processCommandLineArguments(Main.java:151)\n",
      "\tat ca.pfv.spmf.gui.Main.main(Main.java:64)\n",
      "PAI 1/1: 0 unique patterns\n",
      "java.lang.NumberFormatException: For input string: \"Y\"\n",
      "\tat java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:67)\n",
      "\tat java.base/java.lang.Integer.parseInt(Integer.java:662)\n",
      "\tat java.base/java.lang.Integer.parseInt(Integer.java:778)\n",
      "\tat ca.pfv.spmf.algorithms.sequentialpatterns.nosep.AlgoNOSEP.readInputFile(AlgoNOSEP.java:809)\n",
      "\tat ca.pfv.spmf.algorithms.sequentialpatterns.nosep.AlgoNOSEP.runAlgorithm(AlgoNOSEP.java:878)\n",
      "\tat ca.pfv.spmf.algorithmmanager.descriptions.DescriptionAlgoNOSEP.runAlgorithm(DescriptionAlgoNOSEP.java:52)\n",
      "\tat ca.pfv.spmf.gui.CommandProcessor.runAlgorithm(CommandProcessor.java:265)\n",
      "\tat ca.pfv.spmf.gui.Main.processCommandLineArguments(Main.java:151)\n",
      "\tat ca.pfv.spmf.gui.Main.main(Main.java:64)\n",
      "CPAI 1/1: 0 unique patterns\n",
      "NPAI 1/1: island too large\n",
      "--------------------------\n",
      "Genome 38/53: NC_017374\n",
      "No PAI sequences\n",
      "CPAI 1/1: island too large\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 39/53: NC_017376\n",
      "No PAI sequences\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 0\n",
      " Total time ~ 25 ms\n",
      " Maximum memory usage : 14.599899291992188 mb\n",
      " Calculation count 0\n",
      "===================================================\n",
      "\n",
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 0\n",
      " Total time ~ 9 ms\n",
      " Maximum memory usage : 10.952568054199219 mb\n",
      " Calculation count 0\n",
      "===================================================\n",
      "\n",
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 22144\n",
      " Total time ~ 97453 ms\n",
      " Maximum memory usage : 173.3035888671875 mb\n",
      " Calculation count 44501\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CPAI 1/1: 307 unique patterns\n",
      "NPAI 1/1: island too large\n",
      "--------------------------\n",
      "Genome 40/53: NC_017367\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 47917\n",
      " Total time ~ 287666 ms\n",
      " Maximum memory usage : 55.83085632324219 mb\n",
      " Calculation count 77450\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "PAI 1/1: 19282 unique patterns\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 22377\n",
      " Total time ~ 116151 ms\n",
      " Maximum memory usage : 191.81851959228516 mb\n",
      " Calculation count 44887\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CPAI 1/2: 317 unique patterns\n",
      "CPAI 2/2: island too large\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 41/53: NC_014555\n",
      "No PAI sequences\n",
      "CPAI 1/2: island too large\n",
      "CPAI 2/2: island too large\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 42/53: NC_017379\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/2: island too large\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 20214\n",
      " Total time ~ 90437 ms\n",
      " Maximum memory usage : 189.5379867553711 mb\n",
      " Calculation count 39945\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "NPAI 2/2: 469 unique patterns\n",
      "--------------------------\n",
      "Genome 43/53: NC_017363\n",
      "No PAI sequences\n",
      "No CPAI sequences\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 9298\n",
      " Total time ~ 20304 ms\n",
      " Maximum memory usage : 71.9657974243164 mb\n",
      " Calculation count 18817\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "NPAI 1/1: 5071 unique patterns\n",
      "--------------------------\n",
      "Genome 44/53: NC_011498\n",
      "PAI 1/1: island too large\n",
      "No CPAI sequences\n",
      "NPAI 1/2: island too large\n",
      "NPAI 2/2: island too large\n",
      "--------------------------\n",
      "Genome 45/53: NC_017357\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: island too large\n",
      "--------------------------\n",
      "Genome 46/53: NC_017742\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: island too large\n",
      "--------------------------\n",
      "Genome 47/53: NC_017361\n",
      "No PAI sequences\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 28956\n",
      " Total time ~ 196069 ms\n",
      " Maximum memory usage : 144.9538116455078 mb\n",
      " Calculation count 58126\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "CPAI 1/2: 494 unique patterns\n",
      "CPAI 2/2: island too large\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 9094\n",
      " Total time ~ 17542 ms\n",
      " Maximum memory usage : 214.18203735351562 mb\n",
      " Calculation count 18565\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "NPAI 1/1: 66 unique patterns\n",
      "--------------------------\n",
      "Genome 48/53: NC_017381\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: island too large\n",
      "--------------------------\n",
      "Genome 49/53: NC_017362\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: island too large\n",
      "--------------------------\n",
      "Genome 50/53: NC_021215\n",
      "No PAI sequences\n",
      "CPAI 1/1: island too large\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 51/53: NC_017360\n",
      "No PAI sequences\n",
      "CPAI 1/1: island too large\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 52/53: NC_017365\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">/home/maja/.local/lib/python3.10/site-packages/spmf/spmf.jar\n",
      "=============  NOSEP v2.46 - STATS =============\n",
      " Number of patterns found: 58019\n",
      " Total time ~ 349504 ms\n",
      " Maximum memory usage : 159.26554107666016 mb\n",
      " Calculation count 88015\n",
      "===================================================\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "PAI 1/1: 28634 unique patterns\n",
      "CPAI 1/1: island too large\n",
      "No NPAI sequences\n",
      "--------------------------\n",
      "Genome 53/53: NC_017354\n",
      "No PAI sequences\n",
      "No CPAI sequences\n",
      "NPAI 1/1: island too large\n",
      "--------------------------\n"
     ]
    }
   ],
   "source": [
    "unique_pai_patterns  = {}\n",
    "unique_cpai_patterns = {}\n",
    "unique_npai_patterns = {}\n",
    "\n",
    "k_max = 40\n",
    "logging.info(f'Max pattern length: {k_max}')\n",
    "logging.info('--------------------------')\n",
    "\n",
    "for i, (name, data) in enumerate(h_pylori_data.items()):\n",
    "    logging.info(f'Genome {i+1}/{len(h_pylori_data)}: {name}')\n",
    "    \n",
    "    if len(data['pai_sequences']) != 0:\n",
    "        unique_pai_patterns[name] = {}\n",
    "        genome_sequence = data['genome_sequence']\n",
    "        for j, (pai_sequence, indices) in enumerate(data['pai_sequences']):\n",
    "            \n",
    "            if indices[1] - indices[0] > 40000:\n",
    "                logging.info(f\"PAI {j+1}/{len(data['pai_sequences'])}: island too large\")\n",
    "                continue\n",
    "                \n",
    "            unique_patterns = run_nosep(genome_sequence, pai_sequence, indices[0], indices[1], k_max)\n",
    "\n",
    "            unique_pai_patterns[name]['pai_sequence'] = pai_sequence\n",
    "            unique_pai_patterns[name]['indices'] = indices\n",
    "            unique_pai_patterns[name]['patterns'] = unique_patterns\n",
    "\n",
    "            logging.info(f\"PAI {j+1}/{len(data['pai_sequences'])}: {len(unique_patterns)} unique patterns\")\n",
    "            \n",
    "            os.makedirs(f'h_pylori_paidb/{name}/patterns/nosep/pai/', exist_ok=True)\n",
    "            \n",
    "            with open(f'h_pylori_paidb/{name}/patterns/nosep/pai/pai_{j}_patterns.txt', 'w') as file:\n",
    "                for unique_pattern, frequency in unique_patterns:\n",
    "                    file.write(f'{unique_pattern}: {frequency}\\n')\n",
    "                        \n",
    "    else:\n",
    "        logging.info('No PAI sequences')\n",
    "    \n",
    "    if len(data['cpai_sequences']) != 0:    \n",
    "        unique_cpai_patterns[name] = {}\n",
    "\n",
    "        genome_sequence = data['genome_sequence']\n",
    "        for j, (cpai_sequence, indices) in enumerate(data['cpai_sequences']):\n",
    "            \n",
    "            if indices[1] - indices[0] > 40000:\n",
    "                logging.info(f\"CPAI {j+1}/{len(data['cpai_sequences'])}: island too large\")\n",
    "                continue\n",
    "            \n",
    "            unique_patterns = run_nosep(genome_sequence, cpai_sequence, indices[0], indices[1], k_max)\n",
    "\n",
    "            unique_cpai_patterns[name]['cpai_sequence'] = cpai_sequence\n",
    "            unique_cpai_patterns[name]['indices'] = indices\n",
    "            unique_cpai_patterns[name]['patterns'] = unique_patterns\n",
    "            \n",
    "            logging.info(f\"CPAI {j+1}/{len(data['cpai_sequences'])}: {len(unique_patterns)} unique patterns\")\n",
    "            \n",
    "            os.makedirs(f'h_pylori_paidb/{name}/patterns/nosep/cpai/', exist_ok=True)\n",
    "            \n",
    "            with open(f'h_pylori_paidb/{name}/patterns/nosep/cpai/cpai_{j}_patterns.txt', 'w') as file:\n",
    "                for unique_pattern, frequency in unique_patterns:\n",
    "                    file.write(f'{unique_pattern}: {frequency}\\n')\n",
    "            \n",
    "    else:\n",
    "        logging.info('No CPAI sequences')\n",
    "        \n",
    "    if len(data['npai_sequences']) != 0:\n",
    "        unique_npai_patterns[name] = {}\n",
    "        genome_sequence = data['genome_sequence']\n",
    "        for j, (npai_sequence, indices) in enumerate(data['npai_sequences']):\n",
    "            \n",
    "            if indices[1] - indices[0] > 40000:\n",
    "                logging.info(f\"NPAI {j+1}/{len(data['npai_sequences'])}: island too large\")\n",
    "                continue\n",
    "            \n",
    "            unique_patterns = run_nosep(genome_sequence, npai_sequence, indices[0], indices[1], k_max)\n",
    "\n",
    "            unique_npai_patterns[name]['npai_sequence'] = npai_sequence\n",
    "            unique_npai_patterns[name]['indices'] = indices\n",
    "            unique_npai_patterns[name]['patterns'] = unique_patterns\n",
    "\n",
    "            logging.info(f\"NPAI {j+1}/{len(data['npai_sequences'])}: {len(unique_patterns)} unique patterns\")\n",
    "            \n",
    "            os.makedirs(f'h_pylori_paidb/{name}/patterns/nosep/npai/', exist_ok=True)\n",
    "            \n",
    "            with open(f'h_pylori_paidb/{name}/patterns/nosep/npai/npai_{j}_patterns.txt', 'w') as file:\n",
    "                for unique_pattern, frequency in unique_patterns:\n",
    "                    file.write(f'{unique_pattern}: {frequency}\\n')\n",
    "    \n",
    "    else:\n",
    "        logging.info('No NPAI sequences')\n",
    "    logging.info('--------------------------')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "052f80dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('./json/nosep/pai_patterns.json', 'w') as file:\n",
    "    json.dump(unique_pai_patterns, file)\n",
    "    \n",
    "with open('./json/nosep/cpai_patterns.json', 'w') as file:\n",
    "    json.dump(unique_cpai_patterns, file)\n",
    "    \n",
    "with open('./json/nosep/npai_patterns.json', 'w') as file:\n",
    "    json.dump(unique_npai_patterns, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "207e548c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
